### PR TITLE
pios_video: change suspected bugs

### LINKED ISSUE #1588: Seppuku: OSD Vsync issue can cause OSD crash FC WDOG reboot
* Seen in version (release version or githash): 4721d629
* On flight controller (if applicable): Seppuku

Issue details:
Under the right lighting conditions, OSD looses vsync with camera or switches from NTSC to PAL and disappears.  Four seconds later the FC reboots.  This looks to be similar to the issue we had with the RE1.

I can reproduce this with the sun in the upper right corner of the screen.

EDIT: Add Build Details
RMRC Hellbender 204 V1 frame
Seppuku pre-release test FC/OSD
TBS zero/zero camera
RMRC cricket 200mw VTX
Omnivision 3 lobe RHCP antenna
BrainFPV mPB PDB
Speedex ES 25A   OS125
Spektrum FPV Racing diversity RX
Sunnysky x2204s-16 2300kv motors
DALPROP 5045 triblade
Tattu 1550mah 4S 75C

screen grabs from video:

Just before OSD glitch
justbeforeosdloss

OSD format shift
justbeforeosdloss2

OSD Loss
osdloss

WDOG after FC rebooted in flight
wdog

Two videos demonstrating the issue
https://youtu.be/EOTPI6e50Zo

https://youtu.be/BubItLA9I7U

### PR DESCRIPTION
**Pull request by mlyle**
I am unable to fully verify this-- or test this at all-- this moment, but I've noticed several suspect things in pios_video in code audit, that could do really bad things if hsync is toggling too much (for no good reason)

Also, it's possible to filter the video hsync, so let's try enabling it.

Fixes #1588

### CODE REVIEW

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
Main commit comment:

1) ~~we re-enable the line counter for unclear reasons during the frame.
We certainly don't **use** the line counter during a frame.~~
2) It looks like we don't clear the line counter interrupt flag in all
cases.  Surely this is bad if we get a excessively high number of lines?
3) Add a PIOS_VIDEO_INPUT_FILTER thingy that lets us require a minimum
hsync width to trigger.
commit a07ceb6372dd6ec9184c2e26ff35ff13fc774c66

File: flight/PiOS/STM32F4xx/pios_video.c [hunk 1]
```diff
@@ -367,15 +367,34 @@ void PIOS_Video_Init(const struct pios_video_cfg *cfg)
...
+		tmpccer |= (uint16_t)(TIM_CCER_CC2P);
+	}
+#endif
 
-	uint16_t tmpccer = cfg->hsync_capture.timer->CCER;
-	tmpccer &= (uint16_t)~(TIM_CCER_CC1NP);
-	tmpccer |= (uint16_t)(TIM_CCER_CC1P);
-	cfg->hsync_capture.timer->CCER = tmpccer;
+#ifdef PIOS_VIDEO_INPUT_FILTER
+	if (cfg->hsync_capture.timer_chan == TIM_Channel_1) {
```
Comment (tracernz) [inline flight/PiOS/STM32F4xx/pios_video.c:408]: Would appreciate a small comment explaining what IC1F is/does so people don't need to refer to the STM32 ref. manual. :)

File: flight/PiOS/STM32F4xx/pios_video.c [hunk 2]
```diff
@@ -367,15 +367,34 @@ void PIOS_Video_Init(const struct pios_video_cfg *cfg)
...
+		tmpccer &= ((uint16_t)~TIM_CCMR1_IC1F);
+		tmpccer |= 6 << 4;
+		/* 6 = Fdts/4, N=6.  APB1=42MHz, so 6 cycles at 10.5Mhz...
+		 * require a steady value for half a microsecond
+		 */
```
Comment (tracernz) [inline flight/PiOS/STM32F4xx/pios_video.c]: This is good, but doesn't explain what/how Fdts or N mean/are derived.

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
Have confirmed that on this version, the OSD works.  @litterbug-- can you see if these artifacts fix the crash scenario?

It looks like to me if csync toggled 65535 times during a frame before, we'd have entered an infinite loop.  Now there's multiple lines of defense against that (input filtering to not respond to glitches; and the actual bug in the interrupt handler is fixed).

@hshred -- do you want this fix for the original brain?  If so, can you please certify it on brain hw?  (Instead, the line handler removal PR that I have open that I've asked you to look at repeatedly would fix this issue, too, if you'd prefer to go down that route...)

**Comment by mlyle** [comment]
LOL, correction to names.  @mluessi, @litterbugs, see above please

**Comment by LitterBugs** [comment]
I'll give it a test tonight.  :+1:

**Comment by mluessi** [comment]
@mlyle I can test it on brain if it fixes the problem on seppuku.

By "PR that I have open that I've asked you to look at repeatedly" do you mean #1523? I had tested in on brain and it wasn't working, so I'm not sure what you mean.

Edit: Just noticed you actually made changes to #1523. I can test it again if you think it's better and it also fixes this problem.

**Comment by mlyle** [comment]
@mluessi Yes, I've asked you about #1523 four (now five) times since the changes.  It works on Seppuku/PAL.  You failed before on Brain/NTSC.  I need to get out NTSC stuff to test still.

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
#1523 is now a part of this.  A timer is removed and line tracking is now much simpler / easier to analyze on OSD targets.  Confirmed functional on seppuku/PAL

**Review by tracernz — APPROVED** [review]
This looks like a nice improvement. :+1: Ready for merge after some testing. I would suggest combinations that cover all of these at least once:
- [ ] Brain OG
- [x] Seppuku
- [x] NTSC
- [x] PAL

Also, might be nice to fix up `graphics_hight_real` in pios_video_quadspi.c too. :wink:

**Comment by mlyle** [comment]
Seppuku PAL was tested by me; NTSC by both pug & litterbug

**Comment by mlyle** [comment]
@mluessi ping

**Comment by dustin** [comment]
I had a few bits of weirdness Monday with my OG Brain in my caipirinha.  It settled down after a few minutes, but was making me just a bit nervous over the ocean.

Here are several frames demonstrating weirdness.  You can also download [all the frames](https://drive.google.com/open?id=0B-RsldXxKyjhTDlfNHR3SlhEN3c) if you want to see more.

image
image
image
image
image
image

**Comment by mlyle** [comment]
"SENSORS" and sync anomalies sure sounds like a cousin to this problem.

**Comment by tracernz** [comment]
@mluessi ping, can you do Brain + NTSC test?

**Comment by LitterBugs** [comment]
@mluessi any reserve of OG Brain FCs that could be distributed for testing?

**Comment by mluessi** [comment]
Seems to work fine on OG Brain / NTSC. 

@dustin did you have sync problems before on this plane or only with this PR?

**Comment by mlyle** [comment]
@mluessi that was --without-- this PR

**Comment by mluessi** [comment]
OK, that makes sense :)

**Comment by mlyle** [comment]
@mluessi -- thanks for testing.  are you cool with this merging, then?

**Comment by mluessi** [comment]
Yes, looks good. Thanks for the fixes

**Pull request merged**